### DISCUSSION TITLE
Astroid Framework 3.1.0-RC is ready for testing

### DISCUSSION THREAD

**Discussion by sonvnn** [templaza/astroid-framework#722] [Announcements]
Hi all,

I am pleased to announce that Astroid 3.1.0-RC1 is ready for testing. You can see changelogs here https://github.com/templaza/astroid-framework/releases/tag/v3.1.0-rc1

You can download Astroid 3.1.0-rc1 here
https://github.com/templaza/astroid-framework/releases/download/v3.1.0-rc1/astroid-framework-3.1.0-rc1.zip

To test sub-layouts, see this guide: https://docs.astroidframe.work/layout-settings/sub-layouts

To test Article layout builder, see this guide: https://docs.astroidframe.work/article-blog/article-layout-builder

### Note: 
You can test Article layout builder with Two, at the moment or override `templates/{yourtemplate}/html/com_content/article/default.php` with https://github.com/templaza/astroid-framework/blob/two-template/html/com_content/article/default.php

Another way replace content of `templates/{yourtemplate}/html/com_content/article/default.php`
with

> defined('_JEXEC') or die;
> 
> use Joomla\CMS\Factory;
> 
> // Astroid Article/Blog
> if (!isset($astroidArticle)) {
>     $astroidArticle = new Astroid\Article($this->item, false);
> }
> $astroidArticle->renderLayout();

Please let me know if you find any problems. Thanks.

Best Regards,
Sonny

**Comment by sonvnn** [comment]
Astroid 3.1.0-rc2 is released. You can download here https://github.com/templaza/astroid-framework/releases/tag/v3.1.0-rc2

**Comment by sonvnn** [comment]
@Chacapamac @Joomlaplates @rinka88 @paurray @Banuk1616 Please help me take a look on it. Thanks :)

**Reply by rinka88** [reply]
For some reason they disappeared in the “Menu Assignment” widget sections, except for the component area

**Reply by sonvnn** [reply]
@rinka88 WOW!! Thank you very much. It's my mistake. I forgot add condition when $mode is "article_layout_data". I fixed it on https://github.com/templaza/astroid-framework/commit/67472721eec38e280d86764fc6fc96a418e874df#diff-11c60f205d216658a47c436955810cfe7e566bb0b9dd75a88365db3fe2f4afc5

Please help me check it again.

**Reply by rinka88** [reply]
how to install the fix? Which link can I use to download the installation package?

**Reply by sonvnn** [reply]
You can download v3.1.0-rc3 in release page

**Comment by sonvnn** [comment]
@deltapapa01 @Dnsftl @darekmer Please help me take a look on it. Thanks :)

**Reply by rinka88** [reply]
I tested it a couple of times, but didn’t find any particular remark. Perfect solution. I would like to add that it would be better if they added the ability to add a selection for each field separately, in the section of the article layout designer.

**Reply by sonvnn** [reply]
Yes, Article layout builder allows you create many diffirent layouts. You can assign them to each Joomla category.
For example:
1. Create 2 Joomla Categories named: Blog and Portfolio
2. Go to templates -> Create 2 article layouts name: Blog and Portfolio
3. Configure Blog as normally blog, configure Portfolio layout as a 2 column with Image and many widget included.
4. Go back to 2 Joomla Categories. Assign Blog layout to Blog category and assign Portfolio layout to Portfolio category.
5. Create 2 Menus for Blog and Portfolio. 
6. Go to front-end and you can see it works :)

**Reply by rinka88** [reply]
I meant fields and a group of fields that can be used when creating materials. You call it using article_events. I just meant to use each field separately

**Reply by sonvnn** [reply]
Unfortulately, I think I can not do anything in this case. Astroid is just a front-end application. And I don't want to interfere deeply with the core back-end of Joomla.

To achieve the requirements as you suggest. It will require coordination from other complex components and system plugins. We will have to create other extensions and it is beyond the scope of Astroid.

**Reply by rinka88** [reply]
In the Yootheme pro template, when using dynamic content, you can select fields

**Reply by sonvnn** [reply]
@rinka88 Can you send me a Yootheme pro version for testing? May be I can learn something about it. Thank you.

**Reply by rinka88** [reply]
> Я имел в виду поля и группу полей, которые можно использовать при создании материалов. Вы вызываете это с помощью article_events. Я просто имел в виду использовать каждое поле отдельно

@vtheod I also asked about fields here)

**Reply by rinka88** [reply]
I don't currently have any Yootheme subscriptions. But there is a video on Youtube where they use this mechanization. 
https://www.youtube.com/watch?v=A3sdDs4Njl8&list=PLrqT0WH0HPdNUQm6HZlvOfKxXemeOEtLf&index=6
Or maybe someone has a subscription to it

**Comment by joomlaplates-de** [comment]
Hi Sony, for now everything is working.
BTW did you changed the template for using article layout builder or sub layouts? For example the html overrides or anything?
Thanks, Peter

**Reply by sonvnn** [reply]
Hi @Joomlaplates 

In the meantime you can test Article Layout Builder with Two. You can still override templates as usual. Even with your existing templates. The system will still work as it is doing.

If you want to override specific elements like `article_introtext` or `article_fulltext` you can refer to this tutorial https://docs.astroidframe.work/developers/build-or-override-a-widget

Don't hesitate to ask any questions if you have any doubts. I will discuss with you to clarify the issue.

Best Regards,
Sonny

**Comment by nidzo80** [comment]
This is absolutely amazing! Article Layout Builder gives you so many options. Just put imagination to work. And Sublayouts! Brilliant.
This is totaly new aproach. We can already build website without using aditional page builder.  
Thank you Sonny!
I tested and did not notice any bugs.

**Reply by sonvnn** [reply]
Thank you! Keep up your great work. Someones found some bugs. Let's invest more time for testing. 3.1.0 is a big step, so we need check it carefully, together. Thank you for your great work :)

**Comment by sonvnn** [comment]
Astroid Framework v3.1.0 RC3 is ready for testing. You can download it here
https://github.com/templaza/astroid-framework/releases/tag/v3.1.0-rc3

**Comment by Dnsftl** [comment]
3.1.0-RC tested

After setting up sublayout and articles layout builder, the article "page-header" on all other pages disappeared and cannot be recovered.

I have no idea whether this became the case after installing 3.1.0-RC or whether it occurred after setting up the sublayout and articles layout builder. 

Regards,

Dennis

**Reply by sonvnn** [reply]
Opp! My mistake :)) I forgot this element. I will add it on next Monday. Thank you for report bug. Please ignore it and continue for testing.

**Reply by Dnsftl** [reply]
No problem, we are all human, it is not a release candidate for nothing. ;-)

**Reply by sonvnn** [reply]
@Dnsftl I fixed this issue at https://github.com/templaza/astroid-framework/pull/696/commits/a1aaa92cf47846dc19ab474b53a7f438bb2e794d Please help me check it.

**Reply by sonvnn** [reply]
You can download RC4 here https://github.com/templaza/astroid-framework/releases/tag/v3.1.0-rc4

**Comment by Banuk1616** [comment]
Hi Sonny, first of all thank you for your work on the Astroid Framework. During a first quick test of the RC2 I couldn't find anything conspicuous, but I haven't tested everything yet. I first have to familiarise myself with the new functions/possibilities and see exactly how they work and how to use them. If I find some time today I will install RC3 and test it further.
Best regards
Olli

**Reply by sonvnn** [reply]
Yeap! Take your time. No hurry! 3.1.0 is a big step and we need check it carefully.

**Comment by Banuk1616** [comment]
I have just realised something that I had previously overlooked. With the versions RC1, RC2 and RC3 I get the following warnings on the local development environment of my website:

Warning: Undefined variable $assocParam in D:\Bearsampp\vhosts\dev.XXXXX.local\libraries\astroid\framework\elements\article_infoblock\article_infoblock.php on line 25

If I go back to Astroid version 3.0.15, this warning is gone again.

Best Regards 
Olli

**Reply by sonvnn** [reply]
Thank you very much! I forgot $assocParam. I will add it to article_infoblock on the next Monday. Have a nice weekend my friend :)

**Reply by sonvnn** [reply]
@Banuk1616 I fixed this issue at https://github.com/templaza/astroid-framework/pull/696/commits/a1aaa92cf47846dc19ab474b53a7f438bb2e794d please help me check it. Thanks.

**Reply by Banuk1616** [reply]
@sonvnn 
I can confirm that the warning is now gone.

**Comment by joomlaplates-de** [comment]
I couldnt get it to work. Can I send you the Url and admin access?

**Reply by sonvnn** [reply]
Can you detail the issue? I will check it first.

**Reply by joomlaplates-de** [reply]
I have created a new article layout and it doesnt show on frontpage. Blog is assigned to the new article layout.

**Reply by sonvnn** [reply]
Can you send me your admin acc? I will check it on the next Monday.

**Reply by joomlaplates-de** [reply]
Hi Sony, I did it.
I need to replace the file in html folder > com_content > article / default.php
I just copied the complete html folder from template two to my template folder.
Thanks Sony, have a nice weekend.
PS: This function of the new Article Layout lift up the Astroid to the sky... WOW

**Reply by joomlaplates-de** [reply]
BTW is it possible to get the documentation from Gitbook in a html file or pdf?
I want to do a translation to German language.
Thanks

**Reply by sonvnn** [reply]
Unfortunately, PDF is not available in the current plan.

**Comment by deltapapa01** [comment]
Hi so far it looks good for me, Joomla 5.1.2, Astroid Framework 3.1.0 RC3 and Template Zero (3.0.12) and Template Two (1.0.5)
During my test I discovered an error: If I create a SubLayout (for example a simple text), then after saving the SubLayout my Topbar Section is no longer displayed in the frontend.
I then did some further research and found that the default module position Topbar-1 and Topbar-2 are no longer present.
I can see that when I go to the Social / Social Profiles section under Astroid Templatemanager, the Module Position field is empty. I can select other module positions there (in the dropdown), but Topbar-1 and 2 are no longer present in the dropdown field.
If I delete the SubLayout, my Topbar section disappears.
Thanks for the great template Sonny, I will continue testing :-)

**Comment by deltapapa01** [comment]
So, I have now tested the Article Layout Builder. I think it's a wonderful idea, it allows you to create a lot of different layouts very quickly without having to work with overrides.
During testing, however, I noticed that many elements have no content at all. For example, if I select Typography -> Text, I can edit Element Title, Custom Class and Custom ID, and that's it. No content, no formatting, nothing. The same goes for Image and various other elements (actually all elements). What works is Heading, where a heading is displayed in the front end.
Otherwise the frontent always remains empty, unless I select Article Whole Text, in which case the page is displayed as usual.
Small example. These are my settings:
![image](https://github.com/user-attachments/assets/9e2caeb0-1eef-4c5a-9639-db3aadfc9646)

This is what i get:
![image](https://github.com/user-attachments/assets/a2c63c57-8090-4a02-ab76-e6e7d826a44d)

No text, no author info, no picture.
Either the elements do not make sense at this point, or the editing of the content is missing.

WBR,
deltapapa

**Reply by joomlaplates-de** [reply]
Same here

**Reply by sonvnn** [reply]
@deltapapa01 @Joomlaplates To add Widget content in Article follow steps below:

1. Go to administrator >> Content >> Articles >> Edit an article
2. Switch to Widgets tab >> You can see widgets below

![ScreenShot Tool -20240713165550](https://github.com/user-attachments/assets/554b3e23-7145-469f-a9c1-3f4dfac0ef98)

3. Click edit a widget and input your content

![ScreenShot Tool -20240713165634](https://github.com/user-attachments/assets/79d270cf-30b7-4454-8d5f-fc91d4f9b1c7)

Hope it is clear. If it still does not work, please send me your administrator via sonlv@templaza.com
I will help you check it.

Best Regards,
Sonny

**Reply by sonvnn** [reply]
In the front-end you can see the result

![ScreenShot Tool -20240713171243](https://github.com/user-attachments/assets/7c73a7a0-0195-446d-9582-a7f52ac6c7a1)

**Comment by Dnsftl** [comment]
When Joomla (5.1.2.) is in dark mode, there is a minor cosmetic issue in version 3.1.0 RC3.

Steps to Reproduce the Problem:

1. Go to Articles.
2. Open an article with the Article Builder Layout activated.
3. Go to Widgets.
4. Click on Widgets in this article.
5. Choose "Edit".
6. In the menu that appears, select "Select Categories", "Info Block Options" or "FontFamily".

Problem: The dropdown menus are transparent, exposing the underlying settings and making it difficult to make a selection.

Additional Issues:

With the following options:

1. Font Size
2. Letter Spacing
3. Line Height
4. FontFamily (twice)
5. Grid Options

The hover and font color are very dark, which makes it difficult to read the text.

Regards

Dennis

**Reply by sonvnn** [reply]
@Dnsftl Thank you very much! There are many things to check. I fixed this issue at https://github.com/templaza/astroid-framework/pull/696/commits/a1651d9400b3e3626845a0dc60234294a292d049

**Reply by sonvnn** [reply]
You can download RC4 here https://github.com/templaza/astroid-framework/releases/tag/v3.1.0-rc4

**Reply by Dnsftl** [reply]
It now works as expected, (Y)

**Comment by deltapapa01** [comment]
Hello Sonny,
maybe I misunderstood. 
As an example: I have a blog with various articles in my Joomla. I do this via a category, the block view and the individual articles.
But now I would like to have an overview of my customer projects. To do this, I would like to create a category, insert various articles there and fill these articles with an introductory image and text content. This way, every user can easily create new articles. However, I would like to display the customer projects a little differently to the other block posts: For example, I would like to display the introductory image on the left-hand side with 4 columns and the text content of the article with 8 columns. And I thought I could use the article layout builder for this instead of working with a blog override as usual.
Unfortunately, this doesn't work for me at all: https://p603133.mittwaldserver.info/cms/portfolio
If I have understood you correctly, each widget must be configured individually in each post, which is impossible with several hundred posts in a blog.
I really like the idea of the layout builder, but I still don't know how to implement it.
Thank you very much for all your work,

**Reply by sonvnn** [reply]
@deltapapa01 Article Layout is work for Article detail only. It will not work on Category Blog view. In your case you can do follow steps below. 

1. Create 2 categories for Blog and Portfolio
2. Create a layout for Blog or use Default setting. 
3. Create a layout for Portfolio and assign to Portfolio category
4. Input your article content
5. Go to front-end and check it.

Best Regards,
Sonny

**Reply by deltapapa01** [reply]
Hi Sonny,
That's exactly what I did. This is what the layout looks like in the Layout Builder article:
![image](https://github.com/user-attachments/assets/a237f49e-816a-451d-8b08-a51cd917abfd)

And that's what I get:
![image](https://github.com/user-attachments/assets/33d43ae3-2aeb-47f5-b94a-c047bcfef823)

Unfortunately, the image and text are not displayed next to each other. I have the feeling that the layout is somehow not taken over...

**Reply by sonvnn** [reply]
I see it is category view not article view.. . As I mention above, Article layout works with article view only. 

Let's enable show title and click to an article to show detail, then you can see it work.

**Reply by deltapapa01** [reply]
Hi Sonny,
thank you very much, but I still don't understand where the advantages of the article builder lie.
If I have to configure the widgets in every single article anyway, I might as well do it in the post, right?
In addition, you write above that I should create a category and the articles in it will then be displayed according to the article layout builder. But how am I supposed to display the category in Joomla if it doesn't work as a blog view?

As I said, I would have loved to test the feature, but I don't understand the principle behind it at all. If I do it as described in the documentation, the result is a blank page, as I have already described above.

**Reply by sonvnn** [reply]
Basically, Article Layout helps admins customize their article layouts more simply instead of technical intervention through templates. Moreover, you can create more creative articles for each article category.

As you know, displaying articles on the category page and the detail page is completely different. On the category page, there is usually only a short introduction about the article. Specific details will be presented on the detail page. That is why I created the Article Layout application for the detailed article.

If you want to customize the layout for blog_items on the category page. This means we will have to create another Blog Category Item layout builder. With the Layout Builder, Sub-Layout Builder, Article Layout Builder, having another Layout Builder editor for Category View is too complicated and unnecessary, which can even confuse users.

At the moment you can customize Blog Category View by Menu >> Blog Menu >> Blog Options > Lead Item Image Options or Intro Item Image Options.

You said you followed the instructions but it only returned a blank page. Could you please email me your admin account via email sonlv@templaza.com I will help you create an example with Article Layout.

**Comment by joomlaplates-de** [comment]
I got a warning like this since update
Warning: Undefined variable $assocParam in 
`xxxx\libraries\astroid\framework\elements\article_infoblock\article_infoblock.php on line 25`

**Reply by sonvnn** [reply]
I fixed this issue at https://github.com/templaza/astroid-framework/pull/696/commits/a1aaa92cf47846dc19ab474b53a7f438bb2e794d please help me check it. Thanks.

**Reply by joomlaplates-de** [reply]
Yes error message is gone with RC4

**Comment by vtheod** [comment]
Hello @sonvnn 
You're doing a great work!
I am trying to build a custom widget. In the xml file, if I enter `<element_type>system</element_type>` I can see it in the layout, but I can't see it in Article Layout Builder. If I enter `<element_type>article</element_type>` in the xml, then it doesn't show up in both.
Do I miss anything?

**Reply by sonvnn** [reply]
If you want to create a widget. You have add element_type is "widget"

**Reply by vtheod** [reply]
You're blazing fast!!!
I did it but it only shows up in the layout, just like the element_type "system"

**Reply by sonvnn** [reply]
@vtheod It's an issue of getTemplate Class. I fixed it at https://github.com/templaza/astroid-framework/commit/7cf85a42cc9482f09bee29cded678b5a87734104
Please help me download v3-branch and retest again.

**Reply by vtheod** [reply]
@sonvnn I just installed the v3-branch, the widget shows up everywhere.
Thank you!
I want to ask you how can I give a predefined value in a widget so that it gets the content I need from the layout builder, not from each article?

**Reply by sonvnn** [reply]
@vtheod At the moment there is no way to do this. But I will think about it. Thanks for your suggestion.
I released v3.1.0 RC5. https://github.com/templaza/astroid-framework/releases/tag/v3.1.0-rc5 Please help me update and check it.

**Reply by vtheod** [reply]
To give you a hint, I was thinking of a widget which would display a custom field by its ID. It would be very helpful in article layout builder

**Reply by sonvnn** [reply]
@vtheod Now you can set predefined value in widgets on Article Layout Builder. https://github.com/templaza/astroid-framework/commit/3b472e2cced5e95f63cbe2fdb2ae62ddbc401709
Please download v3-branch and test again.

**Reply by vtheod** [reply]
Hello @sonvnn 
I tried to create a widget which will load a custom field by id but with no success. Probably I miss something

**Reply by rinka88** [reply]
@vtheod I think if it works out, it will be in future plans)

**Reply by vtheod** [reply]
@rinka88 I don't know. The best I managed to do was to get the data as they are inserted in the db but this way they are useless

**Comment by sonvnn** [comment]
Astroid Framework 3.1.0 RC6 released. I will release 3.1.0 in this weekend. So please help me check everything again.
Thank you for your cooperation. :)

**Reply by Banuk1616** [reply]
Hello, are all templates (zero, one and two) fully compatible with the new features of Astroid Framework 3.1.0 RC6?

**Reply by sonvnn** [reply]
All of them will work with 3.1.0 even old template from other clubs.

**Reply by Banuk1616** [reply]
Hello Sonny, thank you for the information. Then I'm making a mistake somewhere in the Layout Builder article, although I think I've done it exactly as in the instructions. For testing I loaded the default settings, created a category and assigned the article to this category.

But I can't see the widgets in the article editor. The field remains empty.

I will look for my mistake tonight.

**Reply by sonvnn** [reply]
At the moment, Only Two will work with Article Layout Builder. If you want to use on your current template. You have to replace file `ROOT/templates/{your-template}/html/com_content/article/default.php` with
https://github.com/templaza/astroid-framework/blob/two-template/html/com_content/article/default.php

You can copy content and override your template.
